spikes/lk_sensors_imu_lsm6dsox_fusion_calibration/main.cpp

```diff
@@ -0,0 +1,143 @@
+
+#include "drivers/HighResClock.h"
+#include "rtos/ThisThread.h"
+
+#include "CoreI2C.h"
+#include "CoreLSM6DSOX.hpp"
+#include "LogKit.h"
+#include "fusion/Fusion.h"
+
+using namespace std::chrono;
+using namespace leka;
+
+namespace {
+
+namespace imu {
+
+	namespace internal {
+
+		auto drdy_irq = CoreInterruptIn {PinName::SENSOR_IMU_IRQ};
+		auto i2c	  = CoreI2C(PinName::SENSOR_IMU_TH_I2C_SDA, PinName::SENSOR_IMU_TH_I2C_SCL);
+
+	}	// namespace internal
+
+	CoreLSM6DSOX lsm6dsox(internal::i2c, internal::drdy_irq);
+
+}	// namespace imu
+
+namespace fusion {
+
+	constexpr auto kODR_HZ = int {52};
+
+	auto ahrs = FusionAhrs {};
+
+	const FusionAhrsSettings settings = {
+		.gain				   = 0.5F,
+		.accelerationRejection = 10.0F,
+		.magneticRejection	   = 0.0F,
+		.rejectionTimeout	   = static_cast<unsigned int>(5 * kODR_HZ),   // ? # of samples in 5 seconds
+	};
+
+	auto timestamp		   = rtos::Kernel::Clock::now();
+	auto previousTimestamp = rtos::Kernel::Clock::now();
+
+	auto global_offset = FusionOffset {};
+
+	constexpr auto CALIBRATION = bool {true};
+	// constexpr auto CALIBRATION = bool {false};
+
+	void callback(const interface::LSM6DSOX::SensorData &data)
+	{
+		timestamp  = rtos::Kernel::Clock::now();
+		auto ts_us = mbed::HighResClock::now().time_since_epoch().count();
+
+		// ? Acquire latest sensor data
+		auto gyroscope	   = FusionVector {{data.gy.x, data.gy.y, data.gy.z}};
+		auto accelerometer = FusionVector {{data.xl.x, data.xl.y, data.xl.z}};
+
+		if constexpr (CALIBRATION) {
+			// ? Define calibration offsets
+			// ? Data: https://www.dropbox.com/scl/fi/cue7qpb77892rozyjbmcq/2022_01_16-IMU-Calibration_Data.xlsx
+			// const auto gyroscopeOffset = FusionVector {{}};
+			// const auto accelerometerOffset = FusionVector {{}};
+			constexpr auto gyroscopeOffset	   = FusionVector {{0.02544522554, -0.3286247803, 0.3205770357}};
+			constexpr auto accelerometerOffset = FusionVector {{0.006480437024F, -0.01962820621F, 0.003259031049F}};
+
+			// ? Apply calibration offsets
+			gyroscope =
+				FusionCalibrationInertial(gyroscope, FUSION_IDENTITY_MATRIX, FUSION_VECTOR_ONES, gyroscopeOffset);
+			accelerometer = FusionCalibrationInertial(accelerometer, FUSION_IDENTITY_MATRIX, FUSION_VECTOR_ONES,
+													  accelerometerOffset);
+
+			// ? Update gyroscope offset correction algorithm
+			gyroscope = FusionOffsetUpdate(&global_offset, gyroscope);
+		}
+
+		// ? Calculate delta time (in seconds) to account for gyroscope sample clock error
+		auto deltaTime	  = static_cast<float>((timestamp - previousTimestamp).count()) / 1000.F;
+		previousTimestamp = timestamp;
+
+		// ? Update gyroscope AHRS algorithm
+		FusionAhrsUpdateNoMagnetometer(&ahrs, gyroscope, accelerometer, deltaTime);
+
+		// ? Get quaternion
+		const auto quaternion = FusionAhrsGetQuaternion(&ahrs);
+
+		const auto q_w = quaternion.element.w;
+		const auto q_x = quaternion.element.x;
+		const auto q_y = quaternion.element.y;
+		const auto q_z = quaternion.element.z;
```

Comment:
in IMUKit, I agree, they won't be exposed.

for this spike, they are necessary because they are sent to the x-IMU3 software for calibration. It's important to have the low level data.

as of "hard to read and interpret", they are mathematical constructs invented especially to make reasoning about 3D rotations easier. 

People mostly want Yaw/Pitch/Roll but as soon as you start looking under the hood, you need quaternions

An amazing explanation by Ben Eater https://eater.net/quaternions